{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, \"..\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RBP24 dataset preprocessing\n",
    "\n",
    "This dataset originated in GraphProt paper (https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-1-r17?optIn=true). It consists of 24 proteins. For each protein, four files are present: train.positives, train.negatives, ls.positives, ls.negatives. This notebook processes this dataset in dynamic manner. \n",
    "\n",
    "- Distinct folders are created for each split (eg. negatives_ls)\n",
    "- Headers of files are extracted, chromosome coordinates are adjusted for uniform length -> results in .bed files\n",
    "- .bed files are mapped to the reference genome and saved as .tsv files to coresponding folders\n",
    "- Each .tsv file consists of tab delemited columns: chromosome + coordinates and sequence\n",
    "- Secondary structure notation and conservation scores can be added to the .tsv files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "from data_processing.process_rbp24 import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_data_path = Path('../Data/rbp24/raw') #path to raw RBP24 fasta files\n",
    "raw_file_path = Path('../Data/rbp24/raw/ALKBH5_Baltz2012.ls.negatives.fa') #path to raw RBP24 fasta file\n",
    "processed_data_path = Path('../Data/rbp24/processed') #path to output folder\n",
    "ref_fasta_path = '../Data/reference/referecnce_genome/hg19_latest.fa' #path to reference genome\n",
    "conservation_arrays_path = Path('../Data/reference/conservation/arrays') #path conservation scores np.arrays (have to create one first)\n",
    "window_size = 128 #sequence length (coordinate window of .bed files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of files to process: 92\n",
      "...\n",
      "Processing file: ALKBH5_Baltz2012.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/ALKBH5_Baltz2012.tsv\n",
      "File number 1 processed\n",
      "...\n",
      "Processing file: ALKBH5_Baltz2012.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/ALKBH5_Baltz2012.tsv\n",
      "File number 2 processed\n",
      "...\n",
      "Processing file: ALKBH5_Baltz2012.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/ALKBH5_Baltz2012.tsv\n",
      "File number 3 processed\n",
      "...\n",
      "Processing file: ALKBH5_Baltz2012.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/ALKBH5_Baltz2012.tsv\n",
      "File number 4 processed\n",
      "...\n",
      "Processing file: C17ORF85_Baltz2012.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/C17ORF85_Baltz2012.tsv\n",
      "File number 5 processed\n",
      "...\n",
      "Processing file: C17ORF85_Baltz2012.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/C17ORF85_Baltz2012.tsv\n",
      "File number 6 processed\n",
      "...\n",
      "Processing file: C17ORF85_Baltz2012.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/C17ORF85_Baltz2012.tsv\n",
      "File number 7 processed\n",
      "...\n",
      "Processing file: C17ORF85_Baltz2012.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/C17ORF85_Baltz2012.tsv\n",
      "File number 8 processed\n",
      "...\n",
      "Processing file: C22ORF28_Baltz2012.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/C22ORF28_Baltz2012.tsv\n",
      "File number 9 processed\n",
      "...\n",
      "Processing file: C22ORF28_Baltz2012.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/C22ORF28_Baltz2012.tsv\n",
      "File number 10 processed\n",
      "...\n",
      "Processing file: C22ORF28_Baltz2012.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/C22ORF28_Baltz2012.tsv\n",
      "File number 11 processed\n",
      "...\n",
      "Processing file: C22ORF28_Baltz2012.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/C22ORF28_Baltz2012.tsv\n",
      "File number 12 processed\n",
      "...\n",
      "Processing file: CAPRIN1_Baltz2012.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/CAPRIN1_Baltz2012.tsv\n",
      "File number 13 processed\n",
      "...\n",
      "Processing file: CAPRIN1_Baltz2012.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/CAPRIN1_Baltz2012.tsv\n",
      "File number 14 processed\n",
      "...\n",
      "Processing file: CAPRIN1_Baltz2012.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/CAPRIN1_Baltz2012.tsv\n",
      "File number 15 processed\n",
      "...\n",
      "Processing file: CAPRIN1_Baltz2012.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/CAPRIN1_Baltz2012.tsv\n",
      "File number 16 processed\n",
      "...\n",
      "Processing file: CLIPSEQ_AGO2.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/CLIPSEQ_AGO2.tsv\n",
      "File number 17 processed\n",
      "...\n",
      "Processing file: CLIPSEQ_AGO2.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/CLIPSEQ_AGO2.tsv\n",
      "File number 18 processed\n",
      "...\n",
      "Processing file: CLIPSEQ_AGO2.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/CLIPSEQ_AGO2.tsv\n",
      "File number 19 processed\n",
      "...\n",
      "Processing file: CLIPSEQ_AGO2.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/CLIPSEQ_AGO2.tsv\n",
      "File number 20 processed\n",
      "...\n",
      "Processing file: CLIPSEQ_ELAVL1.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/CLIPSEQ_ELAVL1.tsv\n",
      "File number 21 processed\n",
      "...\n",
      "Processing file: CLIPSEQ_ELAVL1.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/CLIPSEQ_ELAVL1.tsv\n",
      "File number 22 processed\n",
      "...\n",
      "Processing file: CLIPSEQ_ELAVL1.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/CLIPSEQ_ELAVL1.tsv\n",
      "File number 23 processed\n",
      "...\n",
      "Processing file: CLIPSEQ_ELAVL1.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/CLIPSEQ_ELAVL1.tsv\n",
      "File number 24 processed\n",
      "...\n",
      "Processing file: CLIPSEQ_SFRS1.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/CLIPSEQ_SFRS1.tsv\n",
      "File number 25 processed\n",
      "...\n",
      "Processing file: CLIPSEQ_SFRS1.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/CLIPSEQ_SFRS1.tsv\n",
      "File number 26 processed\n",
      "...\n",
      "Processing file: CLIPSEQ_SFRS1.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/CLIPSEQ_SFRS1.tsv\n",
      "File number 27 processed\n",
      "...\n",
      "Processing file: CLIPSEQ_SFRS1.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/CLIPSEQ_SFRS1.tsv\n",
      "File number 28 processed\n",
      "...\n",
      "Processing file: ICLIP_HNRNPC.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/ICLIP_HNRNPC.tsv\n",
      "File number 29 processed\n",
      "...\n",
      "Processing file: ICLIP_HNRNPC.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/ICLIP_HNRNPC.tsv\n",
      "File number 30 processed\n",
      "...\n",
      "Processing file: ICLIP_HNRNPC.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/ICLIP_HNRNPC.tsv\n",
      "File number 31 processed\n",
      "...\n",
      "Processing file: ICLIP_HNRNPC.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/ICLIP_HNRNPC.tsv\n",
      "File number 32 processed\n",
      "...\n",
      "Processing file: ICLIP_TDP43.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/ICLIP_TDP43.tsv\n",
      "File number 33 processed\n",
      "...\n",
      "Processing file: ICLIP_TDP43.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/ICLIP_TDP43.tsv\n",
      "File number 34 processed\n",
      "...\n",
      "Processing file: ICLIP_TDP43.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/ICLIP_TDP43.tsv\n",
      "File number 35 processed\n",
      "...\n",
      "Processing file: ICLIP_TDP43.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/ICLIP_TDP43.tsv\n",
      "File number 36 processed\n",
      "...\n",
      "Processing file: ICLIP_TIA1.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/ICLIP_TIA1.tsv\n",
      "File number 37 processed\n",
      "...\n",
      "Processing file: ICLIP_TIA1.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/ICLIP_TIA1.tsv\n",
      "File number 38 processed\n",
      "...\n",
      "Processing file: ICLIP_TIA1.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/ICLIP_TIA1.tsv\n",
      "File number 39 processed\n",
      "...\n",
      "Processing file: ICLIP_TIA1.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/ICLIP_TIA1.tsv\n",
      "File number 40 processed\n",
      "...\n",
      "Processing file: ICLIP_TIAL1.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/ICLIP_TIAL1.tsv\n",
      "File number 41 processed\n",
      "...\n",
      "Processing file: ICLIP_TIAL1.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/ICLIP_TIAL1.tsv\n",
      "File number 42 processed\n",
      "...\n",
      "Processing file: ICLIP_TIAL1.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/ICLIP_TIAL1.tsv\n",
      "File number 43 processed\n",
      "...\n",
      "Processing file: ICLIP_TIAL1.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/ICLIP_TIAL1.tsv\n",
      "File number 44 processed\n",
      "...\n",
      "Processing file: PARCLIP_AGO1234.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/PARCLIP_AGO1234.tsv\n",
      "File number 45 processed\n",
      "...\n",
      "Processing file: PARCLIP_AGO1234.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/PARCLIP_AGO1234.tsv\n",
      "File number 46 processed\n",
      "...\n",
      "Processing file: PARCLIP_AGO1234.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/PARCLIP_AGO1234.tsv\n",
      "File number 47 processed\n",
      "...\n",
      "Processing file: PARCLIP_AGO1234.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/PARCLIP_AGO1234.tsv\n",
      "File number 48 processed\n",
      "...\n",
      "Processing file: PARCLIP_ELAVL1.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/PARCLIP_ELAVL1.tsv\n",
      "File number 49 processed\n",
      "...\n",
      "Processing file: PARCLIP_ELAVL1.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/PARCLIP_ELAVL1.tsv\n",
      "File number 50 processed\n",
      "...\n",
      "Processing file: PARCLIP_ELAVL1.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/PARCLIP_ELAVL1.tsv\n",
      "File number 51 processed\n",
      "...\n",
      "Processing file: PARCLIP_ELAVL1.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/PARCLIP_ELAVL1.tsv\n",
      "File number 52 processed\n",
      "...\n",
      "Processing file: PARCLIP_ELAVL1A.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/PARCLIP_ELAVL1A.tsv\n",
      "File number 53 processed\n",
      "...\n",
      "Processing file: PARCLIP_ELAVL1A.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/PARCLIP_ELAVL1A.tsv\n",
      "File number 54 processed\n",
      "...\n",
      "Processing file: PARCLIP_ELAVL1A.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/PARCLIP_ELAVL1A.tsv\n",
      "File number 55 processed\n",
      "...\n",
      "Processing file: PARCLIP_ELAVL1A.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/PARCLIP_ELAVL1A.tsv\n",
      "File number 56 processed\n",
      "...\n",
      "Processing file: PARCLIP_EWSR1.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/PARCLIP_EWSR1.tsv\n",
      "File number 57 processed\n",
      "...\n",
      "Processing file: PARCLIP_EWSR1.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/PARCLIP_EWSR1.tsv\n",
      "File number 58 processed\n",
      "...\n",
      "Processing file: PARCLIP_EWSR1.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/PARCLIP_EWSR1.tsv\n",
      "File number 59 processed\n",
      "...\n",
      "Processing file: PARCLIP_EWSR1.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/PARCLIP_EWSR1.tsv\n",
      "File number 60 processed\n",
      "...\n",
      "Processing file: PARCLIP_FUS.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/PARCLIP_FUS.tsv\n",
      "File number 61 processed\n",
      "...\n",
      "Processing file: PARCLIP_FUS.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/PARCLIP_FUS.tsv\n",
      "File number 62 processed\n",
      "...\n",
      "Processing file: PARCLIP_FUS.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/PARCLIP_FUS.tsv\n",
      "File number 63 processed\n",
      "...\n",
      "Processing file: PARCLIP_FUS.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/PARCLIP_FUS.tsv\n",
      "File number 64 processed\n",
      "...\n",
      "Processing file: PARCLIP_HUR.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/PARCLIP_HUR.tsv\n",
      "File number 65 processed\n",
      "...\n",
      "Processing file: PARCLIP_HUR.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/PARCLIP_HUR.tsv\n",
      "File number 66 processed\n",
      "...\n",
      "Processing file: PARCLIP_HUR.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/PARCLIP_HUR.tsv\n",
      "File number 67 processed\n",
      "...\n",
      "Processing file: PARCLIP_HUR.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/PARCLIP_HUR.tsv\n",
      "File number 68 processed\n",
      "...\n",
      "Processing file: PARCLIP_IGF2BP123.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/PARCLIP_IGF2BP123.tsv\n",
      "File number 69 processed\n",
      "...\n",
      "Processing file: PARCLIP_IGF2BP123.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/PARCLIP_IGF2BP123.tsv\n",
      "File number 70 processed\n",
      "...\n",
      "Processing file: PARCLIP_IGF2BP123.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/PARCLIP_IGF2BP123.tsv\n",
      "File number 71 processed\n",
      "...\n",
      "Processing file: PARCLIP_IGF2BP123.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/PARCLIP_IGF2BP123.tsv\n",
      "File number 72 processed\n",
      "...\n",
      "Processing file: PARCLIP_MOV10_Sievers.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/PARCLIP_MOV10_Sievers.tsv\n",
      "File number 73 processed\n",
      "...\n",
      "Processing file: PARCLIP_MOV10_Sievers.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/PARCLIP_MOV10_Sievers.tsv\n",
      "File number 74 processed\n",
      "...\n",
      "Processing file: PARCLIP_MOV10_Sievers.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/PARCLIP_MOV10_Sievers.tsv\n",
      "File number 75 processed\n",
      "...\n",
      "Processing file: PARCLIP_MOV10_Sievers.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/PARCLIP_MOV10_Sievers.tsv\n",
      "File number 76 processed\n",
      "...\n",
      "Processing file: PARCLIP_PUM2.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/PARCLIP_PUM2.tsv\n",
      "File number 77 processed\n",
      "...\n",
      "Processing file: PARCLIP_PUM2.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/PARCLIP_PUM2.tsv\n",
      "File number 78 processed\n",
      "...\n",
      "Processing file: PARCLIP_PUM2.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/PARCLIP_PUM2.tsv\n",
      "File number 79 processed\n",
      "...\n",
      "Processing file: PARCLIP_PUM2.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/PARCLIP_PUM2.tsv\n",
      "File number 80 processed\n",
      "...\n",
      "Processing file: PARCLIP_QKI.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/PARCLIP_QKI.tsv\n",
      "File number 81 processed\n",
      "...\n",
      "Processing file: PARCLIP_QKI.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/PARCLIP_QKI.tsv\n",
      "File number 82 processed\n",
      "...\n",
      "Processing file: PARCLIP_QKI.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/PARCLIP_QKI.tsv\n",
      "File number 83 processed\n",
      "...\n",
      "Processing file: PARCLIP_QKI.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/PARCLIP_QKI.tsv\n",
      "File number 84 processed\n",
      "...\n",
      "Processing file: PARCLIP_TAF15.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/PARCLIP_TAF15.tsv\n",
      "File number 85 processed\n",
      "...\n",
      "Processing file: PARCLIP_TAF15.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/PARCLIP_TAF15.tsv\n",
      "File number 86 processed\n",
      "...\n",
      "Processing file: PARCLIP_TAF15.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/PARCLIP_TAF15.tsv\n",
      "File number 87 processed\n",
      "...\n",
      "Processing file: PARCLIP_TAF15.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/PARCLIP_TAF15.tsv\n",
      "File number 88 processed\n",
      "...\n",
      "Processing file: ZC3H7B_Baltz2012.ls.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_ls/ZC3H7B_Baltz2012.tsv\n",
      "File number 89 processed\n",
      "...\n",
      "Processing file: ZC3H7B_Baltz2012.ls.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_ls/ZC3H7B_Baltz2012.tsv\n",
      "File number 90 processed\n",
      "...\n",
      "Processing file: ZC3H7B_Baltz2012.train.negatives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/negatives_train/ZC3H7B_Baltz2012.tsv\n",
      "File number 91 processed\n",
      "...\n",
      "Processing file: ZC3H7B_Baltz2012.train.positives.fa.gz\n",
      "    Creating bed file...\n",
      "    Mapping to reference fasta...\n",
      "    Saving processed protein to: ../Data/rbp24/processed/positives_train/ZC3H7B_Baltz2012.tsv\n",
      "File number 92 processed\n",
      "...\n"
     ]
    }
   ],
   "source": [
    "#Be aware that secondary structure computation and conservation scores mapping might take a while.\n",
    "#Note that if you want to map conservation scores, conservation arrays have to be created first --> run ../src/data_processing/process_wigfix.py\n",
    "#When folder is passed as a raw_data_path, fuction iterates over all files in folder, and processes .fa files\n",
    "#When file is passed, function simply process that file\n",
    "\n",
    "rbp24_preprocessing(raw_data_path=raw_data_path,\n",
    "                    processed_data_path=processed_data_path, \n",
    "                    ref_fasta_path=ref_fasta_path,\n",
    "                    conservation_arrays_path=conservation_arrays_path,\n",
    "                    window_size=window_size,\n",
    "                    compute_secondary_structure=False,\n",
    "                    map_conservation_scores=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence length: 128\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>sequence</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:25553862-25553990()</td>\n",
       "      <td>TTTATGTACATGTATACACACACACACACACAAATACATCCACATC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:36359687-36359815()</td>\n",
       "      <td>GAGGTGCTGGACATCAGGAACATAGATGAGCAGCCCAAGCCCCTCA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:36699097-36699225()</td>\n",
       "      <td>GAGTGAGACTCCATCTCAAAAAAAAAAAAAAGTTGATTTAAGGCTG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:38385555-38385683()</td>\n",
       "      <td>GTGAGCCAAGATCATGCCACTGCACTCCAGCCTGGGCAACACAGCA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:55650559-55650687()</td>\n",
       "      <td>AAATATAATATAAAACTTTTGGGTACAGCAAAAGCAGTGATTAAAG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 chromosome  \\\n",
       "0  chr1:25553862-25553990()   \n",
       "1  chr1:36359687-36359815()   \n",
       "2  chr1:36699097-36699225()   \n",
       "3  chr1:38385555-38385683()   \n",
       "4  chr1:55650559-55650687()   \n",
       "\n",
       "                                            sequence  label  \n",
       "0  TTTATGTACATGTATACACACACACACACACAAATACATCCACATC...      0  \n",
       "1  GAGGTGCTGGACATCAGGAACATAGATGAGCAGCCCAAGCCCCTCA...      0  \n",
       "2  GAGTGAGACTCCATCTCAAAAAAAAAAAAAAGTTGATTTAAGGCTG...      0  \n",
       "3  GTGAGCCAAGATCATGCCACTGCACTCCAGCCTGGGCAACACAGCA...      0  \n",
       "4  AAATATAATATAAAACTTTTGGGTACAGCAAAAGCAGTGATTAAAG...      0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names = [\"chromosome\", \"sequence\", \"label\"] #add secondary structure and/or conservation scores if added to the data\n",
    "\n",
    "rbp_protein = pd.read_csv(\"../Data/rbp24/processed/negatives_ls/ALKBH5_Baltz2012.tsv\",  delimiter=\"\\t\", header=None, names=names)\n",
    "rbp_protein[\"sequence\"] = rbp_protein.sequence.apply(lambda x: x.upper())\n",
    "print(f\"Sequence length: {len(rbp_protein.sequence[0])}\")\n",
    "rbp_protein.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PRISMNET dataset preprocessing\n",
    "\n",
    "This dataset originated in PRISMNET paper (https://www.nature.com/articles/s41422-021-00476-y). It consists of 173 processed proteins. Each protein is stored as .tsv file. Protein files are splitted to train/ls and positives/negatives and stored into folders like in rbp24 processing. Here raw sequences are used, since preprocessing was performend when creating the dataset (see the link above). Conservation score mapping is not possible since I dont have information about reference genome version."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from data_processing.process_prism import *\n",
    "from data_processing.utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_dir_path = Path(\"../Data/prisment/raw_tsv\")\n",
    "out_dir_path = Path(\"../Data/prisment/processed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of files to process: 173\n",
      "Processing file: AARS_K562.tsv\n",
      "File number 1 processed\n",
      "...\n",
      "Processing file: AATF_K562.tsv\n",
      "File number 2 processed\n",
      "...\n",
      "Processing file: ABCF1_K562.tsv\n",
      "File number 3 processed\n",
      "...\n",
      "Processing file: AGGF1_HepG2.tsv\n",
      "File number 4 processed\n",
      "...\n",
      "Processing file: AGO_HEK293.tsv\n",
      "File number 5 processed\n",
      "...\n",
      "Processing file: AKAP1_K562.tsv\n",
      "File number 6 processed\n",
      "...\n",
      "Processing file: AKAP8L_K562.tsv\n",
      "File number 7 processed\n",
      "...\n",
      "Processing file: ALKBH5_HEK293.tsv\n",
      "File number 8 processed\n",
      "...\n",
      "Processing file: APOBEC3C_K562.tsv\n",
      "File number 9 processed\n",
      "...\n",
      "Processing file: AQR_HepG2.tsv\n",
      "File number 10 processed\n",
      "...\n",
      "Processing file: ATXN2_HEK293T.tsv\n",
      "File number 11 processed\n",
      "...\n",
      "Processing file: AUH_K562.tsv\n",
      "File number 12 processed\n",
      "...\n",
      "Processing file: BCCIP_HepG2.tsv\n",
      "File number 13 processed\n",
      "...\n",
      "Processing file: BCLAF1_HepG2.tsv\n",
      "File number 14 processed\n",
      "...\n",
      "Processing file: BUD13_HepG2.tsv\n",
      "File number 15 processed\n",
      "...\n",
      "Processing file: C17ORF85_HEK293.tsv\n",
      "File number 16 processed\n",
      "...\n",
      "Processing file: C22ORF28_HEK293.tsv\n",
      "File number 17 processed\n",
      "...\n",
      "Processing file: CAPRIN1_HEK293.tsv\n",
      "File number 18 processed\n",
      "...\n",
      "Processing file: CDC40_HepG2.tsv\n",
      "File number 19 processed\n",
      "...\n",
      "Processing file: CPEB4_K562.tsv\n",
      "File number 20 processed\n",
      "...\n",
      "Processing file: CPSF1_HEK293.tsv\n",
      "File number 21 processed\n",
      "...\n",
      "Processing file: CPSF2_HEK293.tsv\n",
      "File number 22 processed\n",
      "...\n",
      "Processing file: CPSF3_HEK293.tsv\n",
      "File number 23 processed\n",
      "...\n",
      "Processing file: CPSF4_HEK293.tsv\n",
      "File number 24 processed\n",
      "...\n",
      "Processing file: CPSF6_K562.tsv\n",
      "File number 25 processed\n",
      "...\n",
      "Processing file: CPSF7_HEK293.tsv\n",
      "File number 26 processed\n",
      "...\n",
      "Processing file: CSTF2T_HepG2.tsv\n",
      "File number 27 processed\n",
      "...\n",
      "Processing file: CSTF2_Hela.tsv\n",
      "File number 28 processed\n",
      "...\n",
      "Processing file: DDX21_K562.tsv\n",
      "File number 29 processed\n",
      "...\n",
      "Processing file: DDX24_K562.tsv\n",
      "File number 30 processed\n",
      "...\n",
      "Processing file: DDX3X_K562.tsv\n",
      "File number 31 processed\n",
      "...\n",
      "Processing file: DDX42_K562.tsv\n",
      "File number 32 processed\n",
      "...\n",
      "Processing file: DDX51_K562.tsv\n",
      "File number 33 processed\n",
      "...\n",
      "Processing file: DDX52_K562.tsv\n",
      "File number 34 processed\n",
      "...\n",
      "Processing file: DDX55_HepG2.tsv\n",
      "File number 35 processed\n",
      "...\n",
      "Processing file: DDX59_HepG2.tsv\n",
      "File number 36 processed\n",
      "...\n",
      "Processing file: DDX6_K562.tsv\n",
      "File number 37 processed\n",
      "...\n",
      "Processing file: DGCR8_K562.tsv\n",
      "File number 38 processed\n",
      "...\n",
      "Processing file: DHX30_K562.tsv\n",
      "File number 39 processed\n",
      "...\n",
      "Processing file: DKC1_HepG2.tsv\n",
      "File number 40 processed\n",
      "...\n",
      "Processing file: DROSHA_K562.tsv\n",
      "File number 41 processed\n",
      "...\n",
      "Processing file: EFTUD2_HepG2.tsv\n",
      "File number 42 processed\n",
      "...\n",
      "Processing file: EIF3D_HepG2.tsv\n",
      "File number 43 processed\n",
      "...\n",
      "Processing file: EIF3G_K562.tsv\n",
      "File number 44 processed\n",
      "...\n",
      "Processing file: EIF3H_HepG2.tsv\n",
      "File number 45 processed\n",
      "...\n",
      "Processing file: EIF4A3_Hela.tsv\n",
      "File number 46 processed\n",
      "...\n",
      "Processing file: eIF4AIII_Hela.tsv\n",
      "File number 47 processed\n",
      "...\n",
      "Processing file: EIF4G2_K562.tsv\n",
      "File number 48 processed\n",
      "...\n",
      "Processing file: ELAVL1_Hela.tsv\n",
      "File number 49 processed\n",
      "...\n",
      "Processing file: EWSR1_HEK293.tsv\n",
      "File number 50 processed\n",
      "...\n",
      "Processing file: EXOSC5_HepG2.tsv\n",
      "File number 51 processed\n",
      "...\n",
      "Processing file: FAM120A_K562.tsv\n",
      "File number 52 processed\n",
      "...\n",
      "Processing file: FASTKD2_K562.tsv\n",
      "File number 53 processed\n",
      "...\n",
      "Processing file: FBL_HEK293.tsv\n",
      "File number 54 processed\n",
      "...\n",
      "Processing file: FIP1L1_HEK293.tsv\n",
      "File number 55 processed\n",
      "...\n",
      "Processing file: FKBP4_HepG2.tsv\n",
      "File number 56 processed\n",
      "...\n",
      "Processing file: FMR1_K562.tsv\n",
      "File number 57 processed\n",
      "...\n",
      "Processing file: FTO_K562.tsv\n",
      "File number 58 processed\n",
      "...\n",
      "Processing file: FUS_HEK293.tsv\n",
      "File number 59 processed\n",
      "...\n",
      "Processing file: FXR1_HEK293.tsv\n",
      "File number 60 processed\n",
      "...\n",
      "Processing file: FXR2_K562.tsv\n",
      "File number 61 processed\n",
      "...\n",
      "Processing file: G3BP1_HepG2.tsv\n",
      "File number 62 processed\n",
      "...\n",
      "Processing file: GEMIN5_K562.tsv\n",
      "File number 63 processed\n",
      "...\n",
      "Processing file: GNL3_K562.tsv\n",
      "File number 64 processed\n",
      "...\n",
      "Processing file: GPKOW_K562.tsv\n",
      "File number 65 processed\n",
      "...\n",
      "Processing file: GRWD1_K562.tsv\n",
      "File number 66 processed\n",
      "...\n",
      "Processing file: GTF2F1_K562.tsv\n",
      "File number 67 processed\n",
      "...\n",
      "Processing file: HLTF_K562.tsv\n",
      "File number 68 processed\n",
      "...\n",
      "Processing file: HNRNPA1_K562.tsv\n",
      "File number 69 processed\n",
      "...\n",
      "Processing file: HNRNPC_Hela.tsv\n",
      "File number 70 processed\n",
      "...\n",
      "Processing file: HNRNPD_HEK293.tsv\n",
      "File number 71 processed\n",
      "...\n",
      "Processing file: HNRNPF_HEK293T.tsv\n",
      "File number 72 processed\n",
      "...\n",
      "Processing file: HNRNPK_K562.tsv\n",
      "File number 73 processed\n",
      "...\n",
      "Processing file: HNRNPM_K562.tsv\n",
      "File number 74 processed\n",
      "...\n",
      "Processing file: HNRNPUL1_K562.tsv\n",
      "File number 75 processed\n",
      "...\n",
      "Processing file: HNRNPU_Hela.tsv\n",
      "File number 76 processed\n",
      "...\n",
      "Processing file: IGF2BP1_K562.tsv\n",
      "File number 77 processed\n",
      "...\n",
      "Processing file: IGF2BP2_HEK293.tsv\n",
      "File number 78 processed\n",
      "...\n",
      "Processing file: IGF2BP3_HEK293.tsv\n",
      "File number 79 processed\n",
      "...\n",
      "Processing file: ILF3_K562.tsv\n",
      "File number 80 processed\n",
      "...\n",
      "Processing file: KHDRBS1_K562.tsv\n",
      "File number 81 processed\n",
      "...\n",
      "Processing file: KHSRP_K562.tsv\n",
      "File number 82 processed\n",
      "...\n",
      "Processing file: LARP4_HepG2.tsv\n",
      "File number 83 processed\n",
      "...\n",
      "Processing file: LARP7_K562.tsv\n",
      "File number 84 processed\n",
      "...\n",
      "Processing file: LIN28A_H9.tsv\n",
      "File number 85 processed\n",
      "...\n",
      "Processing file: LIN28B_K562.tsv\n",
      "File number 86 processed\n",
      "...\n",
      "Processing file: LSM11_K562.tsv\n",
      "File number 87 processed\n",
      "...\n",
      "Processing file: METAP2_K562.tsv\n",
      "File number 88 processed\n",
      "...\n",
      "Processing file: METTL14_Hela.tsv\n",
      "File number 89 processed\n",
      "...\n",
      "Processing file: METTL3_Hela.tsv\n",
      "File number 90 processed\n",
      "...\n",
      "Processing file: MOV10_HEK293.tsv\n",
      "File number 91 processed\n",
      "...\n",
      "Processing file: MTPAP_K562.tsv\n",
      "File number 92 processed\n",
      "...\n",
      "Processing file: NCBP2_HepG2.tsv\n",
      "File number 93 processed\n",
      "...\n",
      "Processing file: NIP7_HepG2.tsv\n",
      "File number 94 processed\n",
      "...\n",
      "Processing file: NIPBL_K562.tsv\n",
      "File number 95 processed\n",
      "...\n",
      "Processing file: NKRF_HepG2.tsv\n",
      "File number 96 processed\n",
      "...\n",
      "Processing file: NOL12_HepG2.tsv\n",
      "File number 97 processed\n",
      "...\n",
      "Processing file: NOLC1_K562.tsv\n",
      "File number 98 processed\n",
      "...\n",
      "Processing file: NONO_K562.tsv\n",
      "File number 99 processed\n",
      "...\n",
      "Processing file: NOP56_HEK293.tsv\n",
      "File number 100 processed\n",
      "...\n",
      "Processing file: NOP58_HEK293.tsv\n",
      "File number 101 processed\n",
      "...\n",
      "Processing file: NPM1_K562.tsv\n",
      "File number 102 processed\n",
      "...\n",
      "Processing file: NUDT21_HEK293.tsv\n",
      "File number 103 processed\n",
      "...\n",
      "Processing file: PABPC4_K562.tsv\n",
      "File number 104 processed\n",
      "...\n",
      "Processing file: PABPN1_HepG2.tsv\n",
      "File number 105 processed\n",
      "...\n",
      "Processing file: PCBP1_K562.tsv\n",
      "File number 106 processed\n",
      "...\n",
      "Processing file: PCBP2_HepG2.tsv\n",
      "File number 107 processed\n",
      "...\n",
      "Processing file: PHF6_K562.tsv\n",
      "File number 108 processed\n",
      "...\n",
      "Processing file: PPIG_HepG2.tsv\n",
      "File number 109 processed\n",
      "...\n",
      "Processing file: PRPF4_HepG2.tsv\n",
      "File number 110 processed\n",
      "...\n",
      "Processing file: PRPF8_K562.tsv\n",
      "File number 111 processed\n",
      "...\n",
      "Processing file: PTBP1PTBP2_Hela.tsv\n",
      "File number 112 processed\n",
      "...\n",
      "Processing file: PTBP1_Hela.tsv\n",
      "File number 113 processed\n",
      "...\n",
      "Processing file: PUM1_K562.tsv\n",
      "File number 114 processed\n",
      "...\n",
      "Processing file: PUM2_HEK293.tsv\n",
      "File number 115 processed\n",
      "...\n",
      "Processing file: PUS1_K562.tsv\n",
      "File number 116 processed\n",
      "...\n",
      "Processing file: QKI_HEK293.tsv\n",
      "File number 117 processed\n",
      "...\n",
      "Processing file: RBFOX2_HepG2.tsv\n",
      "File number 118 processed\n",
      "...\n",
      "Processing file: RBM15_K562.tsv\n",
      "File number 119 processed\n",
      "...\n",
      "Processing file: RBM22_K562.tsv\n",
      "File number 120 processed\n",
      "...\n",
      "Processing file: RBM27_K562.tsv\n",
      "File number 121 processed\n",
      "...\n",
      "Processing file: RBPMS_HEK293.tsv\n",
      "File number 122 processed\n",
      "...\n",
      "Processing file: RPS11_K562.tsv\n",
      "File number 123 processed\n",
      "...\n",
      "Processing file: RPS3_K562.tsv\n",
      "File number 124 processed\n",
      "...\n",
      "Processing file: RTCB_HEK293.tsv\n",
      "File number 125 processed\n",
      "...\n",
      "Processing file: SAFB2_K562.tsv\n",
      "File number 126 processed\n",
      "...\n",
      "Processing file: SBDS_K562.tsv\n",
      "File number 127 processed\n",
      "...\n",
      "Processing file: SDAD1_K562.tsv\n",
      "File number 128 processed\n",
      "...\n",
      "Processing file: SERBP1_K562.tsv\n",
      "File number 129 processed\n",
      "...\n",
      "Processing file: SF3A3_HepG2.tsv\n",
      "File number 130 processed\n",
      "...\n",
      "Processing file: SF3B1_K562.tsv\n",
      "File number 131 processed\n",
      "...\n",
      "Processing file: SF3B4_HepG2.tsv\n",
      "File number 132 processed\n",
      "...\n",
      "Processing file: SLBP_K562.tsv\n",
      "File number 133 processed\n",
      "...\n",
      "Processing file: SLTM_K562.tsv\n",
      "File number 134 processed\n",
      "...\n",
      "Processing file: SMNDC1_K562.tsv\n",
      "File number 135 processed\n",
      "...\n",
      "Processing file: SND1_K562.tsv\n",
      "File number 136 processed\n",
      "...\n",
      "Processing file: SRRM4_HEK293T.tsv\n",
      "File number 137 processed\n",
      "...\n",
      "Processing file: SRSF1_HepG2.tsv\n",
      "File number 138 processed\n",
      "...\n",
      "Processing file: SRSF7_K562.tsv\n",
      "File number 139 processed\n",
      "...\n",
      "Processing file: SRSF9_HepG2.tsv\n",
      "File number 140 processed\n",
      "...\n",
      "Processing file: SUB1_HepG2.tsv\n",
      "File number 141 processed\n",
      "...\n",
      "Processing file: SUPV3L1_K562.tsv\n",
      "File number 142 processed\n",
      "...\n",
      "Processing file: TAF15_HEK293.tsv\n",
      "File number 143 processed\n",
      "...\n",
      "Processing file: TARDBP_K562.tsv\n",
      "File number 144 processed\n",
      "...\n",
      "Processing file: TBRG4_K562.tsv\n",
      "File number 145 processed\n",
      "...\n",
      "Processing file: TIA1_Hela.tsv\n",
      "File number 146 processed\n",
      "...\n",
      "Processing file: TIAL1_Hela.tsv\n",
      "File number 147 processed\n",
      "...\n",
      "Processing file: TNRC6A_K562.tsv\n",
      "File number 148 processed\n",
      "...\n",
      "Processing file: TRA2A_K562.tsv\n",
      "File number 149 processed\n",
      "...\n",
      "Processing file: TROVE2_K562.tsv\n",
      "File number 150 processed\n",
      "...\n",
      "Processing file: U2AF1_K562.tsv\n",
      "File number 151 processed\n",
      "...\n",
      "Processing file: U2AF2_Hela.tsv\n",
      "File number 152 processed\n",
      "...\n",
      "Processing file: U2AF65_Hela.tsv\n",
      "File number 153 processed\n",
      "...\n",
      "Processing file: UCHL5_K562.tsv\n",
      "File number 154 processed\n",
      "...\n",
      "Processing file: UPF1_Hela.tsv\n",
      "File number 155 processed\n",
      "...\n",
      "Processing file: UTP18_K562.tsv\n",
      "File number 156 processed\n",
      "...\n",
      "Processing file: UTP3_K562.tsv\n",
      "File number 157 processed\n",
      "...\n",
      "Processing file: WDR33_HEK293.tsv\n",
      "File number 158 processed\n",
      "...\n",
      "Processing file: WDR3_K562.tsv\n",
      "File number 159 processed\n",
      "...\n",
      "Processing file: WDR43_K562.tsv\n",
      "File number 160 processed\n",
      "...\n",
      "Processing file: WRN_K562.tsv\n",
      "File number 161 processed\n",
      "...\n",
      "Processing file: WTAP_Hela.tsv\n",
      "File number 162 processed\n",
      "...\n",
      "Processing file: XRCC6_K562.tsv\n",
      "File number 163 processed\n",
      "...\n",
      "Processing file: XRN2_K562.tsv\n",
      "File number 164 processed\n",
      "...\n",
      "Processing file: YBX3_K562.tsv\n",
      "File number 165 processed\n",
      "...\n",
      "Processing file: YTHDF2_Hela.tsv\n",
      "File number 166 processed\n",
      "...\n",
      "Processing file: YWHAG_K562.tsv\n",
      "File number 167 processed\n",
      "...\n",
      "Processing file: ZC3H11A_HepG2.tsv\n",
      "File number 168 processed\n",
      "...\n",
      "Processing file: ZC3H7B_HEK293.tsv\n",
      "File number 169 processed\n",
      "...\n",
      "Processing file: ZNF622_K562.tsv\n",
      "File number 170 processed\n",
      "...\n",
      "Processing file: ZNF800_K562.tsv\n",
      "File number 171 processed\n",
      "...\n",
      "Processing file: ZRANB2_K562.tsv\n",
      "File number 172 processed\n",
      "...\n"
     ]
    }
   ],
   "source": [
    "prism_preprocessing(in_dir_path=in_dir_path, \n",
    "                    out_dir_path=out_dir_path,\n",
    "                    compute_secondary_structure=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence length: 101\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>sequence</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000406567|1221|1321</td>\n",
       "      <td>ATTATTGTGACCATCCTGCTGCTCCAGAGTGCCTTTCCAGGTTTCC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000415131|489|589</td>\n",
       "      <td>TCAGAATATTTTGTAATGAAAGGATCTAGAAAGCAACTTGGAAGTG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000373237|1936|2036</td>\n",
       "      <td>TAGATTTATTTATTTATTTAGAGACAGGGTCTCACTCTAGCCCAGG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000409913|1687|1787</td>\n",
       "      <td>TCTGAAAACTAGGCGGAGGTTTCCAAATCTCAATTCTTGACTTCTG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000459956|489|589</td>\n",
       "      <td>CCAAGACCAAGAAGAAGAAGGCTGAGGGCACCGTGTTCACCGAGGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  chromosome  \\\n",
       "0  ENST00000406567|1221|1321   \n",
       "1    ENST00000415131|489|589   \n",
       "2  ENST00000373237|1936|2036   \n",
       "3  ENST00000409913|1687|1787   \n",
       "4    ENST00000459956|489|589   \n",
       "\n",
       "                                            sequence  label  \n",
       "0  ATTATTGTGACCATCCTGCTGCTCCAGAGTGCCTTTCCAGGTTTCC...      0  \n",
       "1  TCAGAATATTTTGTAATGAAAGGATCTAGAAAGCAACTTGGAAGTG...      0  \n",
       "2  TAGATTTATTTATTTATTTAGAGACAGGGTCTCACTCTAGCCCAGG...      0  \n",
       "3  TCTGAAAACTAGGCGGAGGTTTCCAAATCTCAATTCTTGACTTCTG...      0  \n",
       "4  CCAAGACCAAGAAGAAGAAGGCTGAGGGCACCGTGTTCACCGAGGA...      0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names = [\"chromosome\", \"sequence\", \"label\"] #add secondary structure and/or conservation scores if added to the data\n",
    "\n",
    "prism_protein = pd.read_csv(\"../Data/prisment/processed/negatives_ls/AARS_K562.tsv\",  delimiter=\"\\t\", header=None, names=names)\n",
    "prism_protein[\"sequence\"] = prism_protein.sequence.apply(lambda x: x.upper())\n",
    "print(f\"Sequence length: {len(prism_protein.sequence[0])}\")\n",
    "prism_protein.head()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "1ce6ef6cae90fa24a5d611089f848696e2b68a24b2c04701dc1cb8d81e2a4fd7"
  },
  "kernelspec": {
   "display_name": "Python 3.7.11 64-bit ('diplomka': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
