{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gzip\n",
    "from Bio import SeqIO   # for reading fasta files\n",
    "from pyensembl import EnsemblRelease   # to get the gene list\n",
    "import pybedtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pyensembl install --release 75 --species homo_sapiens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ENSEMBL_RELEASE = 75 #emseble release for GRCh37 (hg19)\n",
    "DNA_TOPLEVEL_FASTA_PATH = \"/home/mrkvrbl/Diplomka/Data/reference/referecnce_genome/hg19_latest.fa\"\n",
    "OUTPUT_FILE_CSV = \"/home/mrkvrbl/Diplomka/Data/outputs/random_transcripts.csv\"   # where to save them\n",
    "OUTPUT_FILE_BED = \"/home/mrkvrbl/Diplomka/Data/outputs/random_transcripts.bed\"\n",
    "INTRONS = \"/home/mrkvrbl/Diplomka/Data/reference/region_types/introns.bed.gz\"\n",
    "EXONS = \"/home/mrkvrbl/Diplomka/Data/reference/region_types/exons.bed.gz\"\n",
    "UTR5 = \"/home/mrkvrbl/Diplomka/Data/reference/region_types/5utrs.bed.gz\"\n",
    "UTR3 = \"/home/mrkvrbl/Diplomka/Data/reference/region_types/3utrs.bed.gz\"\n",
    "\n",
    "# to generate random sequences\n",
    "N = 90_000    # how many\n",
    "K = 128     # how long\n",
    "\n",
    "CHRS = [str(chr) for chr in range(1,23)] + ['X', 'Y', 'MT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/mrkvrbl/.cache/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/mrkvrbl/.cache/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.ncrna.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "data = EnsemblRelease(ENSEMBL_RELEASE)\n",
    "human_transcripts = data.transcript_ids()\n",
    "\n",
    "transcripts_full_info  = [data.transcript_by_id(transcript) for transcript in human_transcripts]\n",
    "\n",
    "human_transcript_tuples = [(x.transcript_id, x.gene_id, x.biotype, x.contig, x.start, x.end, x.strand, x.coding_sequence) for x in transcripts_full_info if x.contains_start_codon & x.contains_stop_codon]\n",
    "human_transcript_table = pd.DataFrame.from_records(human_transcript_tuples, columns=[\"id\", \"gene_id\", \"biotype\", \"chr\", \"start\", \"end\", \"strand\", \"coding_sequence\"])\n",
    "assert all(human_transcript_table.start <= human_transcript_table.end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>biotype</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>coding_sequence</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>7</td>\n",
       "      <td>127228399</td>\n",
       "      <td>127231759</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGGGCCTCACCGTGTCCGCGCTCTTTTCGCGGATCTTCGGGAAGA...</td>\n",
       "      <td>543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>12</td>\n",
       "      <td>9092961</td>\n",
       "      <td>9102551</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGTTCCCTTTCTACAGCTGCTGGAGGACTGGACTGCTACTACTAC...</td>\n",
       "      <td>834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>ENSG00000173153</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>11</td>\n",
       "      <td>64073050</td>\n",
       "      <td>64084210</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGTCCAGCCAGGTGGTGGGCATTGAGCCTCTCTACATCAAGGCAG...</td>\n",
       "      <td>1272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>ENSG00000004478</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>12</td>\n",
       "      <td>2904119</td>\n",
       "      <td>2913124</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGACAGCCGAGGAGATGAAGGCGACCGAGAGCGGGGCGCAGTCGG...</td>\n",
       "      <td>1380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000001146</td>\n",
       "      <td>ENSG00000003137</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>72356367</td>\n",
       "      <td>72375167</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGCTCTTTGAGGGCTTGGATCTGGTGTCGGCGCTGGCCACCCTCG...</td>\n",
       "      <td>1539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72858</th>\n",
       "      <td>ENST00000610189</td>\n",
       "      <td>ENSG00000198522</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>27851919</td>\n",
       "      <td>27873713</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGGCGGCGTCCGCAGCTGCCGCTGAGCTCCAGGCTTCTGGGGGTC...</td>\n",
       "      <td>1125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72862</th>\n",
       "      <td>ENST00000610205</td>\n",
       "      <td>ENSG00000173464</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>14</td>\n",
       "      <td>21051054</td>\n",
       "      <td>21058417</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGGAGACCTTTCCTCTGCTGCTGCTCAGCCTGGGCCTGGTTCTTG...</td>\n",
       "      <td>600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72869</th>\n",
       "      <td>ENST00000610222</td>\n",
       "      <td>ENSG00000134207</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>114636381</td>\n",
       "      <td>114696541</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGAGCGGAGTGTGGGGGGCCGGCGGGCCTCGGTGCCAGGAGGCGC...</td>\n",
       "      <td>1533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72878</th>\n",
       "      <td>ENST00000610247</td>\n",
       "      <td>ENSG00000121716</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>7</td>\n",
       "      <td>99933737</td>\n",
       "      <td>99965355</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGGGTCGGCCCCTGCTGCTGCCCCTGCTGCTCCTGCTGCAGCCGC...</td>\n",
       "      <td>684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72883</th>\n",
       "      <td>ENST00000610273</td>\n",
       "      <td>ENSG00000146453</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>6</td>\n",
       "      <td>160221301</td>\n",
       "      <td>160241735</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGTTCTGCACCCGAGGACTGCTATTTTTTGCCTTCCTGGCAGGTC...</td>\n",
       "      <td>1563</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56533 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id          gene_id         biotype chr      start  \\\n",
       "0      ENST00000000233  ENSG00000004059  protein_coding   7  127228399   \n",
       "1      ENST00000000412  ENSG00000003056  protein_coding  12    9092961   \n",
       "2      ENST00000000442  ENSG00000173153  protein_coding  11   64073050   \n",
       "3      ENST00000001008  ENSG00000004478  protein_coding  12    2904119   \n",
       "4      ENST00000001146  ENSG00000003137  protein_coding   2   72356367   \n",
       "...                ...              ...             ...  ..        ...   \n",
       "72858  ENST00000610189  ENSG00000198522  protein_coding   2   27851919   \n",
       "72862  ENST00000610205  ENSG00000173464  protein_coding  14   21051054   \n",
       "72869  ENST00000610222  ENSG00000134207  protein_coding   1  114636381   \n",
       "72878  ENST00000610247  ENSG00000121716  protein_coding   7   99933737   \n",
       "72883  ENST00000610273  ENSG00000146453  protein_coding   6  160221301   \n",
       "\n",
       "             end strand                                    coding_sequence  \\\n",
       "0      127231759      +  ATGGGCCTCACCGTGTCCGCGCTCTTTTCGCGGATCTTCGGGAAGA...   \n",
       "1        9102551      -  ATGTTCCCTTTCTACAGCTGCTGGAGGACTGGACTGCTACTACTAC...   \n",
       "2       64084210      +  ATGTCCAGCCAGGTGGTGGGCATTGAGCCTCTCTACATCAAGGCAG...   \n",
       "3        2913124      +  ATGACAGCCGAGGAGATGAAGGCGACCGAGAGCGGGGCGCAGTCGG...   \n",
       "4       72375167      -  ATGCTCTTTGAGGGCTTGGATCTGGTGTCGGCGCTGGCCACCCTCG...   \n",
       "...          ...    ...                                                ...   \n",
       "72858   27873713      +  ATGGCGGCGTCCGCAGCTGCCGCTGAGCTCCAGGCTTCTGGGGGTC...   \n",
       "72862   21058417      -  ATGGAGACCTTTCCTCTGCTGCTGCTCAGCCTGGGCCTGGTTCTTG...   \n",
       "72869  114696541      -  ATGAGCGGAGTGTGGGGGGCCGGCGGGCCTCGGTGCCAGGAGGCGC...   \n",
       "72878   99965355      +  ATGGGTCGGCCCCTGCTGCTGCCCCTGCTGCTCCTGCTGCAGCCGC...   \n",
       "72883  160241735      +  ATGTTCTGCACCCGAGGACTGCTATTTTTTGCCTTCCTGGCAGGTC...   \n",
       "\n",
       "       length  \n",
       "0         543  \n",
       "1         834  \n",
       "2        1272  \n",
       "3        1380  \n",
       "4        1539  \n",
       "...       ...  \n",
       "72858    1125  \n",
       "72862     600  \n",
       "72869    1533  \n",
       "72878     684  \n",
       "72883    1563  \n",
       "\n",
       "[56533 rows x 9 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assert ~human_transcript_table.coding_sequence.str.contains('N').any()\n",
    "human_transcript_table['length'] = human_transcript_table.coding_sequence.apply(len)\n",
    "human_transcript_table = human_transcript_table[human_transcript_table.biotype == \"protein_coding\"]\n",
    "human_transcript_table = human_transcript_table[human_transcript_table.chr.str.fullmatch(pat=r\"[\\w]{1,2}\")]\n",
    "human_transcript_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1     5574\n",
       "19    4054\n",
       "11    3681\n",
       "2     3629\n",
       "17    3445\n",
       "3     3202\n",
       "12    3202\n",
       "6     2836\n",
       "7     2560\n",
       "16    2486\n",
       "5     2406\n",
       "10    2180\n",
       "X     2169\n",
       "4     2134\n",
       "9     2100\n",
       "8     1990\n",
       "14    1938\n",
       "15    1787\n",
       "20    1428\n",
       "22    1278\n",
       "18     897\n",
       "13     796\n",
       "21     634\n",
       "Y      120\n",
       "MT       7\n",
       "Name: chr, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_transcript_table.chr.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_regions(row, selected_regions):\n",
    "    region_start = 0\n",
    "    region_end = region_start + K\n",
    "    end = row.length\n",
    "\n",
    "    while region_end <= end:\n",
    "       \n",
    "        region = pd.Series([row.gene_id, row.biotype, row.chr, row.start + region_start, row.start + region_end, row.strand, row.coding_sequence[region_start:region_end]], \n",
    "                                    index=[\"gene_id\", \"biotype\", \"chr\", \"start\", \"end\", \"strand\", \"sequence\"])\n",
    "        selected_regions = selected_regions.append(region, ignore_index=True)\n",
    "        \n",
    "        region_start = region_end\n",
    "        region_end = region_end + K\n",
    "    \n",
    "    return selected_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>biotype</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>7</td>\n",
       "      <td>127228399</td>\n",
       "      <td>127228527</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGGGCCTCACCGTGTCCGCGCTCTTTTCGCGGATCTTCGGGAAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>7</td>\n",
       "      <td>127228527</td>\n",
       "      <td>127228655</td>\n",
       "      <td>+</td>\n",
       "      <td>CACCACCATCCCAACCATAGGCTTCAATGTAGAAACAGTGGAATAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>7</td>\n",
       "      <td>127228655</td>\n",
       "      <td>127228783</td>\n",
       "      <td>+</td>\n",
       "      <td>AGGGCCTCATCTTTGTGGTGGACAGTAATGACCGGGAGCGGGTCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>7</td>\n",
       "      <td>127228783</td>\n",
       "      <td>127228911</td>\n",
       "      <td>+</td>\n",
       "      <td>GACATGCCCAACGCCATGCCCGTGAGCGAGCTGACTGACAAGCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>12</td>\n",
       "      <td>9092961</td>\n",
       "      <td>9093089</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGTTCCCTTTCTACAGCTGCTGGAGGACTGGACTGCTACTACTAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id         biotype chr      start        end strand  \\\n",
       "0  ENSG00000004059  protein_coding   7  127228399  127228527      +   \n",
       "1  ENSG00000004059  protein_coding   7  127228527  127228655      +   \n",
       "2  ENSG00000004059  protein_coding   7  127228655  127228783      +   \n",
       "3  ENSG00000004059  protein_coding   7  127228783  127228911      +   \n",
       "4  ENSG00000003056  protein_coding  12    9092961    9093089      -   \n",
       "\n",
       "                                            sequence  \n",
       "0  ATGGGCCTCACCGTGTCCGCGCTCTTTTCGCGGATCTTCGGGAAGA...  \n",
       "1  CACCACCATCCCAACCATAGGCTTCAATGTAGAAACAGTGGAATAT...  \n",
       "2  AGGGCCTCATCTTTGTGGTGGACAGTAATGACCGGGAGCGGGTCCA...  \n",
       "3  GACATGCCCAACGCCATGCCCGTGAGCGAGCTGACTGACAAGCTGG...  \n",
       "4  ATGTTCCCTTTCTACAGCTGCTGGAGGACTGGACTGCTACTACTAC...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_regions = pd.DataFrame(columns=[\"gene_id\", \"biotype\", \"chr\", \"start\", \"end\", \"strand\", \"sequence\"])\n",
    "selected_regions = pd.concat(list(human_transcript_table.apply(lambda x: select_regions(x, selected_regions), axis=1)), ignore_index=True)\n",
    "selected_regions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((56533, 9), (664044, 7))"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_transcript_table.shape, selected_regions.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "protein_coding    664044\n",
       "Name: biotype, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_regions.biotype.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(OUTPUT_FILE_BED, \"w\") as outfile:\n",
    "    for i in range(len(selected_regions)):\n",
    "        line = \"chr\" + selected_regions[\"chr\"][i] + \"\\t\" + str(selected_regions['start'][i]) + \"\\t\" + str(selected_regions['end'][i]) + \"\\n\"\n",
    "        outfile.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(OUTPUT_FILE_CSV, \"w\") as outfile:\n",
    "    for i in range(len(selected_regions)):\n",
    "        line = \"chr\" + selected_regions[\"chr\"][i] + \":\" + str(selected_regions['start'][i]) + \"-\" + str(selected_regions['end'][i]) + \"\\t\" + selected_regions['sequence'][i] +\"\\n\"\n",
    "        outfile.write(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Keep only genes with at least one binding site of rbp24 dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "from pathlib import Path\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mrkvrbl/miniconda3/envs/diplomka/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>127228399</td>\n",
       "      <td>127231759</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>9092961</td>\n",
       "      <td>9102551</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>64073050</td>\n",
       "      <td>64084210</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>2904119</td>\n",
       "      <td>2913124</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>72356367</td>\n",
       "      <td>72375167</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56528</th>\n",
       "      <td>chr2</td>\n",
       "      <td>27851919</td>\n",
       "      <td>27873713</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56529</th>\n",
       "      <td>chr14</td>\n",
       "      <td>21051054</td>\n",
       "      <td>21058417</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56530</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114636381</td>\n",
       "      <td>114696541</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56531</th>\n",
       "      <td>chr7</td>\n",
       "      <td>99933737</td>\n",
       "      <td>99965355</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56532</th>\n",
       "      <td>chr6</td>\n",
       "      <td>160221301</td>\n",
       "      <td>160241735</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56533 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr      start        end strand\n",
       "0       chr7  127228399  127231759      +\n",
       "1      chr12    9092961    9102551      -\n",
       "2      chr11   64073050   64084210      +\n",
       "3      chr12    2904119    2913124      +\n",
       "4       chr2   72356367   72375167      -\n",
       "...      ...        ...        ...    ...\n",
       "56528   chr2   27851919   27873713      +\n",
       "56529  chr14   21051054   21058417      -\n",
       "56530   chr1  114636381  114696541      -\n",
       "56531   chr7   99933737   99965355      +\n",
       "56532   chr6  160221301  160241735      +\n",
       "\n",
       "[56533 rows x 4 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_transcript_table_bed = human_transcript_table[[\"chr\", \"start\", \"end\", \"strand\"]]\n",
    "human_transcript_table_bed[\"chr\"] = \"chr\" + human_transcript_table[\"chr\"]\n",
    "human_transcript_table_bed.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_transcript_table_bed.to_csv(\"/home/mrkvrbl/Diplomka/Data/outputs/random_transcripts_rbp24\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "def intersection(a, b):\n",
    "    a = pybedtools.BedTool(a)\n",
    "    b = pybedtools.BedTool(b)\n",
    "\n",
    "    return a.intersect(b,  u=True, wa=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript = \"/home/mrkvrbl/Diplomka/Data/outputs/random_transcripts_rbp24\"\n",
    "rbp24_path = \"/home/mrkvrbl/Diplomka/Data/rbp24/processed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "intersections = []\n",
    "for root, dirs, files in os.walk(rbp24_path):\n",
    "    if root.endswith(\"train\"):\n",
    "        pos_bed_path = root + \"/positives.bed\"\n",
    "        inter = intersection(transcript, pos_bed_path)\n",
    "        inter_df = pd.read_table(inter.fn, names=[\"chr\", \"start\", \"end\", \"strand\"])\n",
    "        intersections.append(inter_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>127228399</td>\n",
       "      <td>127231759</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>9092961</td>\n",
       "      <td>9102551</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr12</td>\n",
       "      <td>2904119</td>\n",
       "      <td>2913124</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7</td>\n",
       "      <td>91741465</td>\n",
       "      <td>91763844</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24742293</td>\n",
       "      <td>24799466</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42330</th>\n",
       "      <td>chr6</td>\n",
       "      <td>83903104</td>\n",
       "      <td>83906382</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42331</th>\n",
       "      <td>chr13</td>\n",
       "      <td>43894987</td>\n",
       "      <td>43935213</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42332</th>\n",
       "      <td>chr12</td>\n",
       "      <td>94656297</td>\n",
       "      <td>94699374</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42333</th>\n",
       "      <td>chr14</td>\n",
       "      <td>58030640</td>\n",
       "      <td>58332780</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42334</th>\n",
       "      <td>chr10</td>\n",
       "      <td>99627889</td>\n",
       "      <td>99631294</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42335 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr      start        end strand\n",
       "0       chr7  127228399  127231759      +\n",
       "1      chr12    9092961    9102551      -\n",
       "2      chr12    2904119    2913124      +\n",
       "3       chr7   91741465   91763844      -\n",
       "4       chr1   24742293   24799466      +\n",
       "...      ...        ...        ...    ...\n",
       "42330   chr6   83903104   83906382      +\n",
       "42331  chr13   43894987   43935213      -\n",
       "42332  chr12   94656297   94699374      +\n",
       "42333  chr14   58030640   58332780      -\n",
       "42334  chr10   99627889   99631294      +\n",
       "\n",
       "[42335 rows x 4 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rbp24_transcripts = pd.concat(intersections, names=[\"chr\", \"start\", \"end\", \"strand\"]).drop_duplicates().reset_index(drop=True)\n",
    "rbp24_transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "rbp24_transcripts.to_csv(\"/home/mrkvrbl/Diplomka/Data/outputs/transcripts_rbp24.bed\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "for each protein, pick random coordinates from dataset above, than map them to refence genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "1ce6ef6cae90fa24a5d611089f848696e2b68a24b2c04701dc1cb8d81e2a4fd7"
  },
  "kernelspec": {
   "display_name": "Python 3.7.11 64-bit ('diplomka': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
